ksql-streams/src/main/java/io/confluent/ksql/execution/streams/timestamp/LoggingTimestampExtractor.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Copyright 2020 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.streams.timestamp;
+
+import static io.confluent.ksql.serde.util.SerdeProcessingLogMessageFactory.deserializationErrorMsg;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.confluent.ksql.logging.processing.ProcessingLogger;
+import java.util.Objects;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.streams.processor.TimestampExtractor;
+
+/**
+ * A wrapper around {@code TimestampExtractor} that can be configured to suppress any
+ * errors and instead returns a negative timestamp (indicating to streams that the message
+ * should be ignored). Additionally, this class ensures that any errors are logged to the
+ * processing log (even the fatal ones) for visibility.
+ */
+public class LoggingTimestampExtractor implements TimestampExtractor {
+
+  private final TimestampExtractor delegate;
+  private final ProcessingLogger logger;
+  private final boolean failOnError;
+
+  public LoggingTimestampExtractor(
+      final TimestampExtractor delegate,
+      final ProcessingLogger logger,
+      final boolean failOnError
+  ) {
+    this.delegate = Objects.requireNonNull(delegate, "delegate");
+    this.logger = Objects.requireNonNull(logger, "logger");
+    this.failOnError = failOnError;
+  }
+
+  @Override
+  public long extract(final ConsumerRecord<Object, Object> record, final long partitionTime) {
+    try {
+      return delegate.extract(record, partitionTime);
+    } catch (final Exception e) {
+      logger.error(deserializationErrorMsg(e, record));
+      if (failOnError) {
+        throw e;
+      }
+      return -1L;
```

Comment:
Will this allow KSQL to display the row with ROWTIME = -1 in case of a failure? 

I'm trying to understand how the `FailOnInvalidTimestamp` will work in this situation. If a -1 is found on the row, then the `FailOnInvalidTimestamp` will throw an exception which this code will catch and log it to the processing log. But it returns a -1 back to stream, so the stream will set it to as ROWTIME, won't it?

Reply:
any negative timestamp causes the event to be ignored by Kafka Streams. That means that if this wraps `FailOnInvalidTimestamp` but you set `ksql.timestamp.throw.on.invalid` to `false` then it will suppress the throw and just ignore the message